### ISSUE TITLE
Multiple issues with discriminated union regarding alias, delayed annotation and vanishing type hints

### ISSUE THREAD

**Issue by jonas-eschle** [pydantic/pydantic#4582]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

I see some inconsistencies with the types, discrimination, alias and delayed type hints.

The results vary, sometimes one model works, sometimes another. The example encompasses two things: We can run it having aliases (as the script) or without, and we can enforce to have delayed type hints and not (change boolean).

There are two models, one that has our annotated union with a discriminator and one with the same type hints but in a "List".

## One observation: vanishing types

When giving a type hint, it seems that the outermost type hint is often not preserved, i.e. when using "List" or "Union" or "Annotated". When checking `__field__[...].type_`, this seems to be lost (prints at the end). This wrecks havoc potentially if the annotation is removed and the Union only is preserved. It still expects a List, but the Annotation may be gone for good.

## alias vs fieldname

The discriminator goes for the alias. That's a bit surprising to me (is that wanted behavior?) but most notably, it's not consistent.
Test: activate the fields,  i.e. uncomment them.

This gets more confusing when delaying the type hint evaluation (it's done by deliberately raising a `NameError` and updating them manually in the end). Then it still  works without the alias, with the alias (delayed and alias), it only works for the one without a list.

## Summary

There are a couple of things that go weird, I could have opened multiple issues but you may can disentangle them better. A real issue is the vanishing type, the outermost one with annotated unions.

I don't really know what the expected behavior should be, but I think it should be at least consistent.

Let me know if you need any more info

## script

Knobs to tune:
`delayed_hints =True/False`

uncomment/comment the fields with `alias="type"`

### Example Code

```Python
from __future__ import annotations

from typing import Union, List, Tuple

import pydantic
from typing_extensions import Annotated, Literal

from pydantic import Field

delayed_hints = True


class BaseModel(pydantic.BaseModel):
    class Config:
        allow_population_by_field_name = True


class BlackCat(BaseModel):
    cat_type: Literal['cat1'] = Field('cat1', alias='type')
    color: str
    name: str


class WhiteCat(BaseModel):
    cat_type: Literal['cat2'] = Field('cat2', alias='type')
    color: str
    name: str


cats = (BlackCat, WhiteCat)


class Types:
    def __init__(self):
        self.types = tuple(cats)
        self.block = True

    @property
    def cats(self):  # that's just a way of preventing the evaluation of type hints in pydantic
        if delayed_hints and self.block:
            raise NameError("Evaluate forward refs later")
        return Annotated[Union[self.types], Field(discriminator='cat_type')]


types = Types()


class Model(BaseModel):
    pet: types.cats
    test: List[int] = [1]


class ModelList(BaseModel):
    pet: List[types.cats]



if delayed_hints:
    types.block = False  # unblocking
    Model.update_forward_refs()
    ModelList.update_forward_refs()
try:
    model = ModelList(pet=[{'cat_type': 'cat2', 'color': 'black', 'name': 'felix'}])
except Exception as err:
    print(f"ModelList errored with {err}")
else:
    print("ModelList worked!")
    print(f"ModelList: pet (should be list with annotations) {ModelList.__fields__['pet'].type_}")
print("\n\n")
try:
    model2 = Model(pet={'cat_type': 'cat2', 'color': 'black', 'name': 'felix'})
except Exception as err:
    print(f"Model errored with {err}")
else:
    print("Model worked")
    print(f"Model: pet: (should be annotated!) {Model.__fields__['pet'].type_}")
    print(f"ModelList: test (should be List) {Model.__fields__['test'].type_}")
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.2
            pydantic compiled: True
                 install path: /home/jonas/mambaforge/envs/zbenchmark37/lib/python3.7/site-packages/pydantic
               python version: 3.7.13 (default, Mar 29 2022, 02:18:16)  [GCC 7.5.0]
                     platform: Linux-5.15.0-47-generic-x86_64-with-debian-bookworm-sid
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [X] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [X] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [X] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
Descriminated unions have changed a lot in V2, and it's too late to mess with them in V1.

Instead of going into the limitations of them in V1, the most productive thing would be to make sure they work correctly in V2.

Maybe you could have a look at https://github.com/pydantic/pydantic-core/blob/main/tests/validators/test_tagged_union.py and confirm that they resolves theses issues and work for your use cases? Obviously the python type hint sugar hasn't been applied yet, but hopefully there's enough here to demonstrate functionality.

**Comment by sydney-runkle** [comment]
As Samuel mentioned, many of the issues related to discriminated unions were fixed in V2. Feel free to file another issue with questions / comments about V2 behavior, but given that this issue pertains to V1, I'm going to closed as not planned.
